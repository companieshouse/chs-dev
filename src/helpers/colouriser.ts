import { cyan, gray, green, red, yellow } from "ansis";

type ColourMap = [RegExp, (str: string) => string][];

const statusColourMap: ColourMap = [
    [/Exited\s\([1-9]|\d{2,}\)/, red],
    [/Exited\s\(0\)/, yellow],
    [/Up[^()]+$/, green],
    [/Up.+\(health:\sstarting./, cyan],
    [/Up.+\(healthy./, green.bold],
    [/Up.+\(unhealthy/, red.bold],
    [/Not\srunning/, gray]
];

const runStatusColourMap: ColourMap = [
    [/Started/, green],
    [/Healthy/, green.bold],
    [/Stopped/, yellow]
];

const stopStatusColourMap: ColourMap = [
    [/Stopped/, yellow],
    [/Removed/, red]
];

const colourise = (colourMap: ColourMap) => (str: string) => {
    const colour = colourMap
        .find(([regex, _]) => regex.test(str));

    if (colour) {
        return colour[1](str);
    } else {
        return str;
    }
};

type SimpleColour = "cyan" | "red" | "bold-red" | "green" | "bold-green" | "yellow" | "gray" | "grey"

export const simpleColouriser = (str: string, colour: SimpleColour) => {
    let colouriseFn: (str: string) => string;

    switch (colour) {
    case "cyan":
        colouriseFn = cyan;
        break;
    case "red":
        colouriseFn = red;
        break;
    case "bold-red":
        colouriseFn = red.bold;
        break;
    case "green":
        colouriseFn = green;
        break;
    case "bold-green":
        colouriseFn = green.bold;
        break;
    case "yellow":
        colouriseFn = yellow;
        break;
    case "gray":
    case "grey":
        colouriseFn = gray;
        break;
    default:
        colouriseFn = (strToColourise: string) => strToColourise;
        break;
    }

    return colouriseFn(str);
};

export const statusColouriser = colourise(statusColourMap);
export const runStatusColouriser = colourise(runStatusColourMap);
export const stopStatusColouriser = colourise(stopStatusColourMap);
