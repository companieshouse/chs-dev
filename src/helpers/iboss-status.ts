/**
 * checks if the CH_IBOSS_TRIAL variable is set in the environment as a boolean value
 *
 * @returns boolean
 */
export const isIbossEnabled: () => boolean = () => {
    const env = process.env.CH_IBOSS_TRIAL?.toLowerCase();
    return !!(env && env !== "false" && env !== "no" && env !== "0");
};
