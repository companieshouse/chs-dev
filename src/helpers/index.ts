/**
 * Checks if a given value is an object and not an array.
 *
 * @param {unknown} T - The value to check.
 * @returns {boolean} - Returns `true` if the value is an object and not an array, otherwise `false`.
 */
export const isObjectFormat = (T: unknown): boolean => {
    return typeof T === "object" && !Array.isArray(T) && T !== null;
};

/**
 * Determines if an object is empty.
 *
 * @param {Record<string, any>} obj - The object to check.
 * @returns {boolean} - Returns `true` if the object has no own enumerable properties, otherwise `false`.
 */
export const isEmptyObject = (obj: Record<string, any>): boolean => {
    return Object.keys(obj).length === 0;
};

/**
 * Checks if there is at least one common element between two arrays.
 *
 * @param {unknown[]} array1 - The first array to compare.
 * @param {unknown[]} array2 - The second array to compare.
 * @returns {boolean} - Returns `true` if any element in `array1` exists in `array2`, otherwise `false`.
 */
export const matchExistInArrays = (array1: unknown[], array2: unknown[]): boolean => {
    return array1.some(item => array2.includes(item));
};

/**
 * Finds and returns elements from the parentArray that do not exist in the childArray.
 *
 * @param {unknown[]} parentArray - The source array to check.
 * @param {unknown[]} childArray - The reference array to compare against.
 * @returns {any[]} - An array containing elements from `parentArray` that are not present in `childArray`.
 */
export const findUniqueItemsInParentArray = (parentArray: unknown[], childArray: unknown[]): any[] => {
    const childSet = new Set(childArray);
    return parentArray.filter(item => !childSet.has(item));
};
