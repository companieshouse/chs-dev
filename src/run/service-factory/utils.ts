import NewServiceSpec from "../../model/NewServiceSpec.js";

/**
 * Apply the value to the newServiceSpec at the selector. The selector can
 * reference a nested object using dot notation. For example, ".configuration.docker-chs-development.module"
 * but not using array notation (i.e. .configuration["docker-chs-development"]["module"]).
 * @param selector Jq like selector to apply the value to
 * @param newServiceSpec The newServiceSpec to apply the value to
 * @param newValue The value to apply
 * @returns The newServiceSpec with the value applied
 */
export const applySelector = (newServiceSpec: Partial<NewServiceSpec>, selector: string, newValue: any) => {
    const nestedKeys = extractValidNestedKeys(selector);

    if (nestedKeys.length > 1) {
        const newServiceSpecSubAttribute = getNextLevelObject(newServiceSpec, nestedKeys[0]);

        return {
            ...newServiceSpec,
            [nestedKeys[0]]: applySelector(
                newServiceSpecSubAttribute,
                `.${nestedKeys.slice(1).join(".")}`,
                newValue
            )
        };
    } else {
        return {
            ...newServiceSpec,
            [nestedKeys[0]]: newValue
        };
    }
};

/**
 * Lookup a value in the newServiceSpec at the selector. The selector can
 * reference a nested object using dot notation. For example, ".configuration.docker-chs-development.module"
 * but not using array notation (i.e. .configuration["docker-chs-development"]["module"]).
 * @param newServiceSpec NewServiceSpec to lookup the value in
 * @param selector Jq like selector to lookup the value at
 * @returns The value at the selector
 */
export const get = (newServiceSpec: Partial<NewServiceSpec>, selector: string) => {
    const nestedKeys = extractValidNestedKeys(selector);

    if (nestedKeys.length > 1) {
        const newServiceSpecSubAttribute = getNextLevelObject(newServiceSpec, nestedKeys[0]);

        return get(newServiceSpecSubAttribute, `.${nestedKeys.slice(1).join(".")}`);
    } else {
        return newServiceSpec[nestedKeys[0]];
    }
};

const isSelectorValid = (selector: string) => {
    return /^\.([A-Za-z0-9_-]+\.?)+$/.test(selector);
};

const extractValidNestedKeys = (selector: string) => {
    if (!isSelectorValid(selector)) {
        throw new Error("Invalid selector");
    }

    return selector.split(".").slice(1);
};

const getNextLevelObject = (newServiceSpec: Partial<NewServiceSpec>, nestedKey: string) => {
    return newServiceSpec[nestedKey] || {};
};
