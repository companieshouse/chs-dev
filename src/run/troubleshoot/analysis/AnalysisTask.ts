import Config from "../../../model/Config.js";
import { Inventory } from "../../../state/inventory.js";
import { StateManager } from "../../../state/state-manager.js";

export type TroubleshootAnalysisTaskContext = {
    inventory: Inventory,
    stateManager: StateManager,
    config: Config,
}

/**
 * Describes the level of failure - which will describe how the troubleshooter
 * will respond to the failure
 */
export enum AnalysisFailureLevel {

    /**
     * An issue which could cause issues for the environment and so should be
     * investigated further
     */
    WARN,
    /**
     * A critical error which must be addressed and is going to be responsible
     * for issues within the environment.
     */
    FAIL,
    /**
     * Will output a message to the user but will not cause the troubleshoot
     * analyses to fail
     */
    INFO
}

export type AnalysisIssue = {
    title: string,
    description: string,
    suggestions: string[],
    documentationLinks: string[],
}

export type AnalysisTaskOutcome = {
    headline: string,

    level?: AnalysisFailureLevel

    issues: AnalysisIssue[],

    isSuccess(): boolean
}

/**
 * A task which performs an analysis of the state which produces an outcome
 */
export type AnalysisTask = {

    analyse(context: TroubleshootAnalysisTaskContext): Promise<AnalysisTaskOutcome>;
}

export default AnalysisTask;
