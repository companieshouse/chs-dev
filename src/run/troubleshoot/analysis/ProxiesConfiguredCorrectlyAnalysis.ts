import { DockerSettings, fetchDockerSettings } from "../../../helpers/docker-settings-store.js";
import { isOnVpn, isWebProxyHostSet } from "../../../helpers/vpn-check.js";
import BaseAnalysis from "./AbstractBaseAnalysis.js";
import AnalysisOutcome from "./AnalysisOutcome.js";
import { AnalysisIssue, TroubleshootAnalysisTaskContext } from "./AnalysisTask.js";

const ANALYSIS_HEADLINE = "Check user's device proxy configuration";

const WEB_PROXY_SUGGESTION = [
    "Run echo $CH_PROXY_HOST to check web proxy value on device"
];

const DOCKER_PROXY_CONFIGURATION_SUGGESTIONS = [
    "Run 'chproxyon' to set ProxyHttpMode for docker",
    "'ProxyHttpMode' property in docker settings-store.json file must be set to 'manual'",
    "'ProxyHTTPMode' property in docker settings-store.json file must be set to 'manual'",
    "Run echo $HTTPS_PROXY to check web proxy value on device",
    "'OverrideProxyHTTP' property in docker settings-store.json file must correctly match 'HTTPS_PROXY' value",
    "'OverrideProxyHTTPS' property in docker settings-store.json file must correctly match 'HTTPS_PROXY' value"
];

/**
 * An analysis task which checks whether the proxy configuration for the user device are configured correctly.
 */

export default class ProxiesConfiguredCorrectlyAnalysis extends BaseAnalysis {

    async analyse (context: TroubleshootAnalysisTaskContext): Promise<AnalysisOutcome> {

        const vpnIssues = this.checkCHProxyConfig();
        const dockerIssues = this.checkDockerProxiesConfig();

        const issues: AnalysisIssue[] = [vpnIssues, dockerIssues].filter((issue): issue is AnalysisIssue => issue !== undefined);

        return this.createOutcomeFrom(ANALYSIS_HEADLINE, issues, "Fail");
    }

    private checkCHProxyConfig (): AnalysisIssue | undefined {
        if (isWebProxyHostSet()) {
            if (!isOnVpn()) {
                return {
                    title: "CH_PROXY_HOST ping unsuccessful",
                    description: `Ping on webproxy not successful`,
                    suggestions: WEB_PROXY_SUGGESTION,
                    documentationLinks: []
                };
            }
        } else {
            return {
                title: "CH_PROXY_HOST env not set",
                description: `CH_PROXY_HOST value missing in env.`,
                suggestions: WEB_PROXY_SUGGESTION,
                documentationLinks: []
            };
        }
    }

    private checkDockerProxiesConfig (): AnalysisIssue | undefined {
        const httpProxy = process.env.HTTPS_PROXY || process.env.https_proxy;
        const result = fetchDockerSettings();
        const { OverrideProxyHTTP, OverrideProxyHTTPS, ProxyHTTPMode, ProxyHttpMode } = result as DockerSettings;
        const issues = result as AnalysisIssue;

        if (!issues.title) {
            if (!(OverrideProxyHTTP === httpProxy && OverrideProxyHTTPS === httpProxy &&
                ProxyHTTPMode === "manual" && ProxyHttpMode === "manual")) {
                return {
                    title: "Docker proxy settings invalid",
                    description: `Docker proxy settings properties not configured correctly`,
                    suggestions: DOCKER_PROXY_CONFIGURATION_SUGGESTIONS,
                    documentationLinks: []
                };
            }
        } else {
            return issues;
        }

    }

}
