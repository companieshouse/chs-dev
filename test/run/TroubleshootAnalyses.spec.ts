import { expect, jest } from "@jest/globals";
import TroubleshootAnalyses from "../../src/run/TroubleshootAnalyses";
import { modules, services } from "../utils/data";
import { Inventory } from "../../src/state/inventory";
import { StateManager } from "../../src/state/state-manager";
import AnalysisTask, { AnalysisFailureLevel } from "../../src/run/troubleshoot/analysis/AnalysisTask";
import { simpleColouriser as simpleColouriserMock } from "../../src/helpers/colouriser";
import fs from "fs";
import { isIbossEnabled } from "../../src/helpers/iboss-status.js";
import ProxiesConfiguredCorrectlyAnalysis from "../../src/run/troubleshoot/analysis/ProxiesConfiguredCorrectlyAnalysis.js";

const analysisTaskMockOne = {
    analyse: jest.fn()
};

const analysisTaskMockTwo = {
    analyse: jest.fn()
};

const analysisTaskMockThree = {
    analyse: jest.fn()
};

const analysisTasks: AnalysisTask[] = [
    analysisTaskMockOne as AnalysisTask,
    analysisTaskMockTwo as AnalysisTask,
    analysisTaskMockThree as AnalysisTask
];

const inventoryMock = {
    services,
    modules
} as Inventory;

const stateManagerMock = {

    snapshot: {
        modules: [
            "module-one"
        ],
        services: [
            "service-six"
        ],
        servicesWithLiveUpdate: [
            "service-two",
            "service-ten",
            "service-nine"
        ],
        excludedServices: [
            "service-four"
        ]
    }
} as StateManager;

const loggerMock = {
    log: jest.fn()
};

jest.mock("../../src/helpers/colouriser");

jest.mock("../../src/helpers/iboss-status.js", () => {
    return {
        isIbossEnabled: jest.fn()
    };
});

describe("TroubleshootAnalyses", () => {

    const writeFileSyncSpy = jest.spyOn(fs, "writeFileSync");
    const config = {
        projectPath: "/home/user/docker",
        projectName: "docker",
        env: {}
    };

    const defaultMockBehaviour = () => {
        // @ts-expect-error
        simpleColouriserMock.mockImplementation((str: string, _) => str);

        analysisTaskMockOne.analyse.mockResolvedValue({
            headline: "Task one",
            level: AnalysisFailureLevel.WARN,
            issues: [],
            isSuccess: () => true
        } as never);

        analysisTaskMockTwo.analyse.mockResolvedValue({
            headline: "Task two",
            level: AnalysisFailureLevel.FAIL,
            issues: [],
            isSuccess: () => true
        } as never);

        analysisTaskMockThree.analyse.mockResolvedValue({
            headline: "Task three",
            level: AnalysisFailureLevel.INFO,
            issues: [],
            isSuccess: () => true
        } as never);

        writeFileSyncSpy.mockReturnValue(undefined);
    };

    describe("Default behaviour", () => {
        let troubleshootAnalyses: TroubleshootAnalyses;

        const emptyOptions = {};

        beforeEach(() => {
            jest.resetAllMocks();

            troubleshootAnalyses = new TroubleshootAnalyses(
                analysisTasks,
                inventoryMock,
                stateManagerMock,
                config,
                loggerMock
            );

            defaultMockBehaviour();
        });

        it("calls each task", async () => {
            await troubleshootAnalyses.perform(emptyOptions);

            expect(analysisTaskMockOne.analyse).toHaveBeenCalledWith({
                inventory: inventoryMock,
                config,
                stateManager: stateManagerMock
            });

            expect(analysisTaskMockTwo.analyse).toHaveBeenCalledWith({
                inventory: inventoryMock,
                config,
                stateManager: stateManagerMock
            });

            expect(analysisTaskMockThree.analyse).toHaveBeenCalledWith({
                inventory: inventoryMock,
                config,
                stateManager: stateManagerMock
            });
        });

        it("resolves to success outcome when successful", async () => {
            const result = await troubleshootAnalyses.perform(emptyOptions);

            expect(result.success).toBe(true);
        });

        it("resolves to be success outcome when all informational", async () => {
            analysisTaskMockOne.analyse.mockResolvedValue({
                level: AnalysisFailureLevel.INFO,
                issues: [{
                    title: "Missing Dockerfile for service in Development mode without builder label",
                    description: "Missing Dockerfile for service service-ten",
                    suggestions: [
                        "Provide a valid value for the builder label",
                        "Create a Dockerfile"
                    ],
                    documentationLinks: [
                        "troubleshooting-remedies/correctly-setup-repository-builder.md",
                        "troubleshooting-remedies/correctly-setup-builder.md"
                    ]
                }],
                isSuccess: () => false
            } as never);

            const result = await troubleshootAnalyses.perform(emptyOptions);

            expect(result.success).toBe(true);

        });

        it("resolves to unsuccessful outcome when unsuccessful", async () => {
            analysisTaskMockOne.analyse.mockResolvedValue({
                level: AnalysisFailureLevel.WARN,
                issues: [{
                    title: "Missing Dockerfile for service in Development mode without builder label",
                    description: "Missing Dockerfile for service service-ten",
                    suggestions: [
                        "Provide a valid value for the builder label",
                        "Create a Dockerfile"
                    ],
                    documentationLinks: [
                        "troubleshooting-remedies/correctly-setup-repository-builder.md",
                        "troubleshooting-remedies/correctly-setup-builder.md"
                    ]
                }],
                isSuccess: () => false
            } as never);

            const result = await troubleshootAnalyses.perform(emptyOptions);

            expect(result.success).toBe(false);
        });

        it("resolves to unsuccessful outcome when unsuccessful and one info", async () => {
            analysisTaskMockOne.analyse.mockResolvedValue({
                level: AnalysisFailureLevel.WARN,
                issues: [{
                    title: "Missing Dockerfile for service in Development mode without builder label",
                    description: "Missing Dockerfile for service service-ten",
                    suggestions: [
                        "Provide a valid value for the builder label",
                        "Create a Dockerfile"
                    ],
                    documentationLinks: [
                        "troubleshooting-remedies/correctly-setup-repository-builder.md",
                        "troubleshooting-remedies/correctly-setup-builder.md"
                    ]
                }],
                isSuccess: () => false
            } as never);

            analysisTaskMockTwo.analyse.mockResolvedValue({
                level: AnalysisFailureLevel.INFO,
                issues: [{
                    title: "Missing Dockerfile for service in Development mode without builder label",
                    description: "Missing Dockerfile for service service-ten",
                    suggestions: [
                        "Provide a valid value for the builder label",
                        "Create a Dockerfile"
                    ],
                    documentationLinks: [
                        "troubleshooting-remedies/correctly-setup-repository-builder.md",
                        "troubleshooting-remedies/correctly-setup-builder.md"
                    ]
                }],
                isSuccess: () => false
            } as never);

            const result = await troubleshootAnalyses.perform(emptyOptions);

            expect(result.success).toBe(false);
        });

        it("logs a successful message when there are no issues", async () => {
            await troubleshootAnalyses.perform(emptyOptions);

            expect(loggerMock.log).toHaveBeenCalledWith("SUCCESS! No issues found with your docker environment");
            expect(loggerMock.log).toHaveBeenCalledWith("If there are still problems with your environment create a report and seek support");
        });

        it("colourises success", async () => {
            await troubleshootAnalyses.perform(emptyOptions);

            expect(simpleColouriserMock).toHaveBeenCalledWith(
                "SUCCESS!", "bold-green"
            );

            expect(simpleColouriserMock).toHaveBeenCalledWith(
                "If there are still problems with your environment create a report and seek support",
                "grey"
            );
        });

        it("does not log successful messages when there are issues", async () => {
            analysisTaskMockOne.analyse.mockResolvedValue({
                level: AnalysisFailureLevel.WARN,
                issues: [{
                    title: "Missing Dockerfile for service in Development mode without builder label",
                    description: "Missing Dockerfile for service service-ten",
                    suggestions: [
                        "Provide a valid value for the builder label",
                        "Create a Dockerfile"
                    ],
                    documentationLinks: [
                        "troubleshooting-remedies/correctly-setup-repository-builder.md",
                        "troubleshooting-remedies/correctly-setup-builder.md"
                    ]
                }],
                isSuccess: () => false
            } as never);

            await troubleshootAnalyses.perform(emptyOptions);

            expect(loggerMock.log).not.toHaveBeenCalledWith("SUCCESS! No issues found with your docker environment");
            expect(loggerMock.log).not.toHaveBeenCalledWith("If there are still problems with your environment create a report and seek support");
        });

        it("logs an error when there are issues", async () => {

            analysisTaskMockOne.analyse.mockResolvedValue({
                level: AnalysisFailureLevel.WARN,
                issues: [{
                    title: "Missing Dockerfile for service in Development mode without builder label",
                    description: "Missing Dockerfile for service service-ten",
                    suggestions: [
                        "Provide a valid value for the builder label",
                        "Create a Dockerfile"
                    ],
                    documentationLinks: [
                        "troubleshooting-remedies/correctly-setup-repository-builder.md",
                        "troubleshooting-remedies/correctly-setup-builder.md"
                    ]

                }],
                isSuccess: () => false
            } as never);

            await troubleshootAnalyses.perform(emptyOptions);

            expect(loggerMock.log).toHaveBeenCalledWith(
                "There were issues found with your environment"
            );
        });

        it("logs the outcome of failed tasks correctly", async () => {
            analysisTaskMockOne.analyse.mockResolvedValue({
                headline: "Check for services in development mode correctly configured",
                level: AnalysisFailureLevel.WARN,
                issues: [{
                    title: "Missing Dockerfile for service in Development mode without builder label",
                    description: "Missing Dockerfile for service service-ten",
                    suggestions: [
                        "Provide a valid value for the builder label",
                        "Create a Dockerfile"
                    ],
                    documentationLinks: [
                        "troubleshooting-remedies/correctly-setup-repository-builder.md",
                        "troubleshooting-remedies/correctly-setup-builder.md"
                    ]

                }, {
                    title: "Incorrect Builder defined for service in development mode",
                    description: "Service: service-nine contains an incorrect builder label: java-11",
                    suggestions: [
                        "Provide a valid value for the builder label from: java, node",
                        "Remove the builder label and Produce a Dockerfile in the source"
                    ],
                    documentationLinks: [
                        "troubleshooting-remedies/correctly-setup-repository-builder.md",
                        "troubleshooting-remedies/correctly-setup-builder.md"
                    ]
                }],
                isSuccess: () => false
            } as never);

            analysisTaskMockTwo.analyse.mockResolvedValue({
                headline: "Incorrect something else",
                level: AnalysisFailureLevel.FAIL,
                issues: [{
                    title: "Something seriously incorrect with your setup",
                    description: "Description about what is seriously wrong",
                    suggestions: [
                        "Provide a valid value for this or that"
                    ],
                    documentationLinks: [
                        "troubleshooting-remedies/correctly-do-the-right-thing.md"
                    ]

                }],
                isSuccess: () => false
            } as never);

            await troubleshootAnalyses.perform(emptyOptions);

            expect(loggerMock.log.mock.calls).toMatchSnapshot();
        });

        it("colours the outcome of failed tasks correctly", async () => {
            analysisTaskMockOne.analyse.mockResolvedValue({
                headline: "Check for services in development mode correctly configured",
                level: AnalysisFailureLevel.WARN,
                issues: [{
                    title: "Missing Dockerfile for service in Development mode without builder label",
                    description: "Missing Dockerfile for service service-ten",
                    suggestions: [
                        "Provide a valid value for the builder label",
                        "Create a Dockerfile"
                    ],
                    documentationLinks: [
                        "troubleshooting-remedies/correctly-setup-repository-builder.md",
                        "troubleshooting-remedies/correctly-setup-builder.md"
                    ]

                }, {
                    title: "Incorrect Builder defined for service in development mode",
                    description: "Service: service-nine contains an incorrect builder label: java-11",
                    suggestions: [
                        "Provide a valid value for the builder label from: java, node",
                        "Remove the builder label and Produce a Dockerfile in the source"
                    ],
                    documentationLinks: [
                        "troubleshooting-remedies/correctly-setup-repository-builder.md",
                        "troubleshooting-remedies/correctly-setup-builder.md"
                    ]
                }],
                isSuccess: () => false
            } as never);

            analysisTaskMockTwo.analyse.mockResolvedValue({
                headline: "Incorrect something else",
                level: AnalysisFailureLevel.FAIL,
                issues: [{
                    title: "Something seriously incorrect with your setup",
                    description: "Description about what is seriously wrong",
                    suggestions: [
                        "Provide a valid value for this or that"
                    ],
                    documentationLinks: [
                        "troubleshooting-remedies/correctly-do-the-right-thing.md"
                    ]

                }],
                isSuccess: () => false
            } as never);

            analysisTaskMockThree.analyse.mockResolvedValue({
                headline: "Info about something else that worth looking at",
                level: AnalysisFailureLevel.INFO,
                issues: [{
                    title: "Something could be incorrect with your setup",
                    description: "Description about what is could be wrong",
                    suggestions: [
                        "Provide a valid value for this or that"
                    ],
                    documentationLinks: [
                        "troubleshooting-remedies/correctly-do-the-right-thing.md"
                    ]

                }],
                isSuccess: () => false
            } as never);

            await troubleshootAnalyses.perform(emptyOptions);

            expect(simpleColouriserMock).toHaveBeenCalledWith(
                "CRITICAL", "bold-red"
            );

            expect(simpleColouriserMock).toHaveBeenCalledWith(
                "WARNING", "yellow"
            );

            expect(simpleColouriserMock).toHaveBeenCalledWith(
                "QUERY", "cyan"
            );
        });
    });

    describe("quiet", () => {

        let troubleshootAnalyses: TroubleshootAnalyses;

        const quietOptions = {
            quiet: true
        };

        beforeEach(() => {
            jest.resetAllMocks();

            troubleshootAnalyses = new TroubleshootAnalyses(
                analysisTasks,
                inventoryMock,
                stateManagerMock,
                config,
                loggerMock
            );

            defaultMockBehaviour();
        });

        it("does not log when there are no issues", async () => {
            await troubleshootAnalyses.perform(quietOptions);

            expect(loggerMock.log).not.toHaveBeenCalled();
        });

        it("does not log when there are are issues", async () => {
            analysisTaskMockOne.analyse.mockResolvedValue({
                headline: "Check for services in development mode correctly configured",
                level: AnalysisFailureLevel.WARN,
                issues: [{
                    title: "Missing Dockerfile for service in Development mode without builder label",
                    description: "Missing Dockerfile for service service-ten",
                    suggestions: [
                        "Provide a valid value for the builder label",
                        "Create a Dockerfile"
                    ],
                    documentationLinks: [
                        "troubleshooting-remedies/correctly-setup-repository-builder.md",
                        "troubleshooting-remedies/correctly-setup-builder.md"
                    ]

                }, {
                    title: "Incorrect Builder defined for service in development mode",
                    description: "Service: service-nine contains an incorrect builder label: java-11",
                    suggestions: [
                        "Provide a valid value for the builder label from: java, node",
                        "Remove the builder label and Produce a Dockerfile in the source"
                    ],
                    documentationLinks: [
                        "troubleshooting-remedies/correctly-setup-repository-builder.md",
                        "troubleshooting-remedies/correctly-setup-builder.md"
                    ]
                }],
                isSuccess: () => false
            } as never);

            analysisTaskMockTwo.analyse.mockResolvedValue({
                headline: "Incorrect something else",
                level: AnalysisFailureLevel.FAIL,
                issues: [{
                    title: "Something seriously incorrect with your setup",
                    description: "Description about what is seriously wrong",
                    suggestions: [
                        "Provide a valid value for this or that"
                    ],
                    documentationLinks: [
                        "troubleshooting-remedies/correctly-do-the-right-thing.md"
                    ]

                }],
                isSuccess: () => false
            } as never);

            await troubleshootAnalyses.perform(quietOptions);

            expect(loggerMock.log).not.toHaveBeenCalled();
        });
    });

    describe("fileOut", () => {

        let troubleshootAnalyses: TroubleshootAnalyses;

        const fileOutOptions = {
            fileOut: "/home/user/analyses.json"
        };

        beforeEach(() => {
            jest.resetAllMocks();

            troubleshootAnalyses = new TroubleshootAnalyses(
                analysisTasks,
                inventoryMock,
                stateManagerMock,
                config,
                loggerMock
            );

            defaultMockBehaviour();
        });

        it("outputs empty analyses with empty issues when no issues", async () => {
            analysisTaskMockTwo.analyse.mockResolvedValue({
                headline: "Check for services in development mode correctly configured",
                level: AnalysisFailureLevel.FAIL,
                issues: [{
                    title: "Missing Dockerfile for service in Development mode without builder label",
                    description: "Missing Dockerfile for service service-ten",
                    suggestions: [
                        "Provide a valid value for the builder label",
                        "Create a Dockerfile"
                    ],
                    documentationLinks: [
                        "troubleshooting-remedies/correctly-setup-repository-builder.md",
                        "troubleshooting-remedies/correctly-setup-builder.md"
                    ]

                }, {
                    title: "Incorrect Builder defined for service in development mode",
                    description: "Service: service-nine contains an incorrect builder label: java-11",
                    suggestions: [
                        "Provide a valid value for the builder label from: java, node",
                        "Remove the builder label and Produce a Dockerfile in the source"
                    ],
                    documentationLinks: [
                        "troubleshooting-remedies/correctly-setup-repository-builder.md",
                        "troubleshooting-remedies/correctly-setup-builder.md"
                    ]
                }],
                isSuccess: () => false
            } as never);

            await troubleshootAnalyses.perform(fileOutOptions);

            expect(writeFileSyncSpy).toHaveBeenCalledWith(
                fileOutOptions.fileOut,
                JSON.stringify({
                    analysisResults: {
                        "Task one": {
                            criticality: "WARNING",
                            issues: []
                        },
                        "Check for services in development mode correctly configured": {
                            criticality: "CRITICAL",
                            issues: [{
                                title: "Missing Dockerfile for service in Development mode without builder label",
                                description: "Missing Dockerfile for service service-ten",
                                suggestions: [
                                    "Provide a valid value for the builder label",
                                    "Create a Dockerfile"
                                ],
                                documentationLinks: [
                                    "troubleshooting-remedies/correctly-setup-repository-builder.md",
                                    "troubleshooting-remedies/correctly-setup-builder.md"
                                ]

                            }, {
                                title: "Incorrect Builder defined for service in development mode",
                                description: "Service: service-nine contains an incorrect builder label: java-11",
                                suggestions: [
                                    "Provide a valid value for the builder label from: java, node",
                                    "Remove the builder label and Produce a Dockerfile in the source"
                                ],
                                documentationLinks: [
                                    "troubleshooting-remedies/correctly-setup-repository-builder.md",
                                    "troubleshooting-remedies/correctly-setup-builder.md"
                                ]
                            }]
                        },
                        "Task three": {
                            criticality: "QUERY",
                            issues: []
                        }
                    }
                })
            );

        });

        it("The ProxiesConfiguredCorrectlyAnalysis task to be filtered out when iboss is enabled", async () => {
            (isIbossEnabled as jest.Mock).mockReturnValue(true);
            const analysisTaskMockFour = new ProxiesConfiguredCorrectlyAnalysis();

            const task = [...analysisTasks, analysisTaskMockFour];

            // eslint-disable-next-line dot-notation
            const filteredTasks = troubleshootAnalyses["handleIbossCheck"](task);

            expect(filteredTasks).not.toContain(analysisTaskMockFour);

        });

        it("The ProxiesConfiguredCorrectlyAnalysis task to be a task when iboss is disabled", async () => {
            (isIbossEnabled as jest.Mock).mockReturnValue(false);
            const analysisTaskMockFour = new ProxiesConfiguredCorrectlyAnalysis();
            const task = [...analysisTasks, analysisTaskMockFour];

            // eslint-disable-next-line dot-notation
            const filteredTasks = troubleshootAnalyses["handleIbossCheck"](task);

            expect(filteredTasks).toContain(analysisTaskMockFour);

        });

        it("can be combined with quiet", async () => {
            await troubleshootAnalyses.perform({ ...fileOutOptions, quiet: true });

            expect(writeFileSyncSpy).toHaveBeenCalledWith(
                fileOutOptions.fileOut,
                expect.anything()
            );

            expect(loggerMock.log).not.toHaveBeenCalled();
        });
    });
});
