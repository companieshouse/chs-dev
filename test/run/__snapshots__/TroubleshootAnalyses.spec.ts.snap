// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TroubleshootAnalyses Default behaviour logs the outcome of failed tasks correctly 1`] = `
[
  [
    "There were issues found with your environment",
  ],
  [
    "WARNING - Check for services in development mode correctly configured:",
  ],
  [
    "	* Missing Dockerfile for service in Development mode without builder label:",
  ],
  [
    "		Missing Dockerfile for service service-ten
",
  ],
  [
    "		Suggestions:",
  ],
  [
    "			- Provide a valid value for the builder label",
  ],
  [
    "			- Create a Dockerfile",
  ],
  [
    "
		Documentation Links:",
  ],
  [
    "			- https://www.github.com/companieshouse/chs-dev/blob/main/docs/troubleshooting-remedies/correctly-setup-repository-builder.md",
  ],
  [
    "			- https://www.github.com/companieshouse/chs-dev/blob/main/docs/troubleshooting-remedies/correctly-setup-builder.md",
  ],
  [
    "	* Incorrect Builder defined for service in development mode:",
  ],
  [
    "		Service: service-nine contains an incorrect builder label: java-11
",
  ],
  [
    "		Suggestions:",
  ],
  [
    "			- Provide a valid value for the builder label from: java, node",
  ],
  [
    "			- Remove the builder label and Produce a Dockerfile in the source",
  ],
  [
    "
		Documentation Links:",
  ],
  [
    "			- https://www.github.com/companieshouse/chs-dev/blob/main/docs/troubleshooting-remedies/correctly-setup-repository-builder.md",
  ],
  [
    "			- https://www.github.com/companieshouse/chs-dev/blob/main/docs/troubleshooting-remedies/correctly-setup-builder.md",
  ],
  [
    "CRITICAL - Incorrect something else:",
  ],
  [
    "	* Something seriously incorrect with your setup:",
  ],
  [
    "		Description about what is seriously wrong
",
  ],
  [
    "		Suggestions:",
  ],
  [
    "			- Provide a valid value for this or that",
  ],
  [
    "
		Documentation Links:",
  ],
  [
    "			- https://www.github.com/companieshouse/chs-dev/blob/main/docs/troubleshooting-remedies/correctly-do-the-right-thing.md",
  ],
]
`;
